File: src/estimation/stochvol_runner.R
#!/usr/bin/env Rscript
# stochvol_runner.R — Single-series stochvol MCMC for the base SV model.
#
# Implements the ASIS (ancillarity-sufficiency interweaving) sampler from
# Kastner & Frühwirth-Schnatter (2014), via the stochvol R package.
#
# Called by stochvol_runner.py via subprocess. Not intended to be run directly.
#
# Usage:
#   Rscript stochvol_runner.R <input_csv> <output_json> <draws> <burnin> <seed1> <seed2>
#
# Arguments:
#   input_csv   — path to CSV file with one column of log-returns (no header)
#   output_json — path to write JSON output
#   draws       — number of posterior draws per chain
#   burnin      — number of burn-in iterations per chain
#   seed1       — RNG seed for chain 1
#   seed2       — RNG seed for chain 2 (used only for R-hat computation)
#
# Output JSON fields:
#   post_mean  — posterior means [mu, phi, sigma_eta]
#   post_sd    — posterior standard deviations [mu, phi, sigma_eta]
#   rhat       — Gelman-Rubin R-hat [mu, phi, sigma_eta], computed across 2 chains
#   samples    — posterior draws from chain 1, shape (draws, 3): [[mu,phi,sigma_eta],...]
#
# Priors (weakly informative, consistent with simulation ranges):
#   mu        ~ Normal(-5, 3)              covers mu in (-10, 0)       [P ~ 0.90]
#   phi       ~ Beta(5, 1.5) on (phi+1)/2  concentrates mass near 1    [P(phi>0.5) ~ 0.61]
#   sigma_eta ~ sqrt(Gamma(0.5, 0.5))      covers sigma_eta in (0.05,1) [P ~ 0.64]
# These cannot be made exactly Uniform (stochvol does not support Uniform priors), but
# are wide enough that the prior has minimal influence on the posterior for T >= 500.

suppressPackageStartupMessages({
  library(stochvol)
  library(jsonlite)
})

args <- commandArgs(trailingOnly = TRUE)
if (length(args) != 6L) {
  stop("Expected 6 arguments: input_csv output_json draws burnin seed1 seed2")
}

input_csv   <- args[1]
output_json <- args[2]
draws_n     <- as.integer(args[3])
burnin_n    <- as.integer(args[4])
seed1       <- as.integer(args[5])
seed2       <- as.integer(args[6])

# ── Load returns ──────────────────────────────────────────────────────────────
y <- read.csv(input_csv, header = FALSE)[[1]]
if (any(!is.finite(y))) stop("Non-finite values found in input returns.")

# ── Prior specification ───────────────────────────────────────────────────────
priors <- specify_priors(
  mu     = sv_normal(mean = -5.0, sd = 3.0),
  phi    = sv_beta(shape1 = 5, shape2 = 1.5),
  sigma2 = sv_gamma(shape = 0.5, rate = 0.5)
)

# ── Gelman-Rubin R-hat for two chains ────────────────────────────────────────
compute_rhat <- function(chain1, chain2) {
  n           <- length(chain1)
  chain_means <- c(mean(chain1), mean(chain2))
  chain_vars  <- c(var(chain1),  var(chain2))
  grand_mean  <- mean(chain_means)
  B           <- n * var(chain_means)           # between-chain variance * n
  W           <- mean(chain_vars)               # within-chain variance
  var_hat     <- (n - 1L) / n * W + B / n
  sqrt(var_hat / W)
}

# ── Run two chains ────────────────────────────────────────────────────────────
set.seed(seed1)
res1 <- tryCatch(
  svsample(y, draws = draws_n, burnin = burnin_n, priorspec = priors, quiet = TRUE),
  error = function(e) stop(paste("stochvol chain 1 failed:", conditionMessage(e)))
)

set.seed(seed2)
res2 <- tryCatch(
  svsample(y, draws = draws_n, burnin = burnin_n, priorspec = priors, quiet = TRUE),
  error = function(e) stop(paste("stochvol chain 2 failed:", conditionMessage(e)))
)

# ── Extract draws: columns mu, phi, sigma (sigma = sigma_eta, not sigma^2) ───
# para() returns an mcmc object — convert to plain numeric matrix for JSON serialisation
d1 <- matrix(para(res1)[, c("mu", "phi", "sigma")], ncol = 3L,
             dimnames = list(NULL, c("mu", "phi", "sigma")))
d2 <- matrix(para(res2)[, c("mu", "phi", "sigma")], ncol = 3L,
             dimnames = list(NULL, c("mu", "phi", "sigma")))

# ── R-hat per parameter ───────────────────────────────────────────────────────
rhat_vals <- c(
  compute_rhat(d1[, "mu"],    d2[, "mu"]),
  compute_rhat(d1[, "phi"],   d2[, "phi"]),
  compute_rhat(d1[, "sigma"], d2[, "sigma"])
)

# ── Posterior summaries (chain 1) ─────────────────────────────────────────────
post_mean <- colMeans(d1)
post_sd   <- apply(d1, 2L, sd)

# ── Write JSON ────────────────────────────────────────────────────────────────
# samples: (draws, 3) matrix — toJSON serialises row-major: [[mu,phi,sig],...]
out <- list(
  post_mean = unname(as.numeric(post_mean)),
  post_sd   = unname(as.numeric(post_sd)),
  rhat      = unname(as.numeric(rhat_vals)),
  samples   = unname(d1)          # matrix (draws x 3), serialised as array of rows
)

write(toJSON(out, digits = 10L, matrix = "rowmajor"), output_json)

File: src/estimation/stochvol_runner.py
"""
Stochvol benchmark runner for the base SV model.

Wraps the stochvol R package via subprocess, implementing the ASIS
(ancillarity-sufficiency interweaving) sampler from Kastner &
Frühwirth-Schnatter (2014).  Output format is identical to mcmc_runner.py
so downstream scripts are backend-agnostic.

Priors are weakly informative and broadly consistent with the simulation
parameter ranges — see stochvol_runner.R for full documentation.

Requirements: R >= 4.0 and the stochvol + jsonlite R packages installed.
Install via:
    brew install r
    Rscript -e "install.packages(c('stochvol','jsonlite'), repos='https://cloud.r-project.org')"
"""

import json
import logging
import subprocess
import tempfile
from dataclasses import dataclass
from pathlib import Path

import numpy as np
from joblib import Parallel, delayed

from .mcmc_runner import MCMCResult   # reuse the same result dataclass

logger = logging.getLogger(__name__)

_R_SCRIPT = Path(__file__).parent / "stochvol_runner.R"


@dataclass
class StochvolConfig:
    """
    Configuration for the stochvol benchmark sampler.

    Attributes:
        draws:       Posterior draws per chain. Two chains are run; R-hat is
                     computed across both, posterior summaries use chain 1.
        burnin:      Burn-in (warm-up) iterations per chain.
        n_jobs:      Parallel worker processes for batch runs.
        random_seed: Base seed. Per-series seeds are (random_seed + i) for
                     chain 1 and (random_seed + 10_000 + i) for chain 2.
    """
    draws:       int = 1000
    burnin:      int = 1000
    n_jobs:      int = 4
    random_seed: int = 42


def run_stochvol_single(
    returns: np.ndarray,
    config:  StochvolConfig,
    seed:    int,
) -> MCMCResult:
    """
    Fit the base SV model to one return series using stochvol (ASIS sampler).

    Runs two chains for Gelman-Rubin R-hat computation. Posterior summaries
    (mean, std, samples) are taken from chain 1.

    Args:
        returns: Observed log-returns, shape (T,).
        config:  StochvolConfig hyperparameters.
        seed:    Base random seed for this series (chain 2 uses seed + 10_000).

    Returns:
        MCMCResult with posterior mean, std, samples (draws, 3), and rhat.
        Column order throughout: [mu, phi, sigma_eta].

    Raises:
        RuntimeError: If the R subprocess fails or times out.
    """
    seed1 = seed
    seed2 = seed + 10_000

    with tempfile.TemporaryDirectory() as tmp:
        tmp_path    = Path(tmp)
        input_csv   = tmp_path / "returns.csv"
        output_json = tmp_path / "result.json"

        # Write returns: one value per line, no header
        np.savetxt(input_csv, returns, delimiter=",", fmt="%.10f")

        cmd = [
            "Rscript", str(_R_SCRIPT),
            str(input_csv),
            str(output_json),
            str(config.draws),
            str(config.burnin),
            str(seed1),
            str(seed2),
        ]

        try:
            proc = subprocess.run(
                cmd,
                capture_output=True,
                text=True,
                timeout=600,   # 10-minute hard cap per series
            )
        except subprocess.TimeoutExpired:
            raise RuntimeError(
                f"stochvol timed out after 600s (seed={seed})"
            )

        if proc.returncode != 0:
            raise RuntimeError(
                f"stochvol R script failed (seed={seed}, returncode={proc.returncode}):\n"
                f"stderr: {proc.stderr}\nstdout: {proc.stdout}"
            )

        with open(output_json) as f:
            result = json.load(f)

    post_mean = np.array(result["post_mean"], dtype=np.float32)  # (3,)
    post_sd   = np.array(result["post_sd"],   dtype=np.float32)  # (3,)
    rhat      = np.array(result["rhat"],      dtype=np.float32)  # (3,)
    samples   = np.array(result["samples"],   dtype=np.float32)  # (draws, 3)

    # Sanity checks
    assert post_mean.shape == (3,), f"Unexpected post_mean shape: {post_mean.shape}"
    assert samples.shape == (config.draws, 3), f"Unexpected samples shape: {samples.shape}"
    if not np.all(np.isfinite(samples)):
        logger.warning("Non-finite values in stochvol samples (seed=%d).", seed)

    if np.any(rhat > 1.1):
        logger.warning(
            "R-hat > 1.1 (seed=%d): mu=%.3f  phi=%.3f  sigma_eta=%.3f",
            seed, rhat[0], rhat[1], rhat[2],
        )

    return MCMCResult(mean=post_mean, std=post_sd, samples=samples, rhat=rhat)


# ---------------------------------------------------------------------------
# Checkpointing helpers — mirror mcmc_runner.py exactly
# ---------------------------------------------------------------------------

def _checkpoint_path(out_path: Path, idx: int) -> Path:
    return out_path / f"series_{idx:04d}.npz"


def _save_checkpoint(result: MCMCResult, idx: int, out_path: Path) -> None:
    p   = _checkpoint_path(out_path, idx)
    tmp = p.with_suffix(".tmp.npz")
    np.savez_compressed(
        tmp,
        mean=result.mean, std=result.std,
        samples=result.samples, rhat=result.rhat,
    )
    tmp.rename(p)   # atomic rename — avoids corrupted files on crash


def _load_checkpoint(idx: int, out_path: Path) -> MCMCResult:
    data = np.load(_checkpoint_path(out_path, idx))
    return MCMCResult(
        mean=data["mean"], std=data["std"],
        samples=data["samples"], rhat=data["rhat"],
    )


def _get_pending_indices(N: int, out_path: Path) -> list[int]:
    return [i for i in range(N) if not _checkpoint_path(out_path, i).exists()]


# ---------------------------------------------------------------------------
# Batch runner
# ---------------------------------------------------------------------------

def run_stochvol_batch(
    dataset,
    config:   StochvolConfig,
    out_path: str | Path,
) -> None:
    """
    Run stochvol on all N series in dataset, with checkpointing and parallelism.

    Output structure mirrors run_mcmc_batch():
        out_path/series_{i:04d}.npz  — per-series checkpoint (crash-safe)
        out_path/results.npz         — assembled batch results

    results.npz arrays:
        means       (N, 3)          float32  — posterior means [mu, phi, sigma_eta]
        stds        (N, 3)          float32  — posterior std devs
        samples     (N, draws, 3)   float32  — posterior draws (chain 1)
        rhats       (N, 3)          float32  — Gelman-Rubin R-hat across 2 chains
        true_params (N, 3)          float32  — ground-truth parameters

    Re-running resumes from the first incomplete series (checkpoint-based).

    Args:
        dataset:  SimulationResult with .returns (N, T) and .params (N, 3).
        config:   StochvolConfig hyperparameters.
        out_path: Directory for per-series checkpoints and final results.npz.
    """
    out_path = Path(out_path)
    out_path.mkdir(parents=True, exist_ok=True)

    N       = dataset.returns.shape[0]
    pending = _get_pending_indices(N, out_path)

    if not pending:
        logger.info("All %d series already completed. Assembling results.", N)
    else:
        logger.info(
            "Running stochvol on %d/%d pending series with %d workers.",
            len(pending), N, config.n_jobs,
        )

        def _run_one(i: int) -> tuple[int, MCMCResult]:
            result = run_stochvol_single(
                dataset.returns[i],
                config,
                seed=config.random_seed + i,
            )
            return i, result

        completed = Parallel(n_jobs=config.n_jobs, backend="multiprocessing")(
            delayed(_run_one)(i) for i in pending
        )

        for i, result in completed:
            _save_checkpoint(result, i, out_path)
            logger.info("Saved series %d/%d", i + 1, N)

    all_results = [_load_checkpoint(i, out_path) for i in range(N)]

    np.savez_compressed(
        out_path / "results.npz",
        means=np.stack([r.mean    for r in all_results]).astype(np.float32),
        stds=np.stack([r.std      for r in all_results]).astype(np.float32),
        samples=np.stack([r.samples for r in all_results]).astype(np.float32),
        rhats=np.stack([r.rhat    for r in all_results]).astype(np.float32),
        true_params=dataset.params[:, :3].astype(np.float32),
    )
    logger.info("Results saved to %s", out_path / "results.npz")
